### PR TITLE
Content based routing

### PR DESCRIPTION
**Pull request by mmihic**
See https://github.com/square/misk/issues/45

Ready for real review. Still some more tests needing around the cases where either the action lacks ``RequestContentType``/``ResponseContentType`` annotations, or the incoming request lacks ``Accept``/``Content-Type`` headers.

### CODE REVIEW

File: misk/src/main/kotlin/misk/util/Checks.kt [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package misk.util
+
+object Checks {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/util/Checks.kt]: Why not just Kotlin check and require ?

Comment (swankjesse) [reply misk/src/main/kotlin/misk/util/Checks.kt]: https://github.com/JetBrains/kotlin/blob/1.2.0/libraries/stdlib/src/kotlin/util/Preconditions.kt#L13

File: misk/src/main/kotlin/misk/web/BoundAction.kt [hunk 2]
```diff
@@ -102,6 +115,19 @@ internal class BoundAction<A : WebAction, R>(
     }
 }
 
+private fun HttpServletRequest.accepts(): List<MediaRange> {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/BoundAction.kt]: Cooool

File: misk/src/main/kotlin/misk/web/extractors/ParameterExtractor.kt [hunk 3]
```diff
@@ -17,9 +18,10 @@ interface ParameterExtractor {
          * Returns an instance of a [ParameterExtractor] that extracts the value for [parameter],
          * or null if the extractor does not apply to [parameter].
          *
-         * See [HeadersParameterExtractorFactory] and [JsonBodyParameterExtractorFactory] for
+         * See [HeadersParameterExtractorFactory] and [RequestBodyParameterExtractor] for
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/extractors/ParameterExtractor.kt]: Dig

File: misk/src/main/kotlin/misk/web/extractors/PathPatternParameterExtractor.kt [hunk 4]
```diff
@@ -13,9 +14,14 @@ import kotlin.reflect.full.findAnnotation
  * and returns it as a [String]. If the parameter name doesn't occur in [pathPattern], returns null.
  */
 object PathPatternParameterExtractorFactory : ParameterExtractor.Factory {
-    override fun create(parameter: KParameter, pathPattern: PathPattern): ParameterExtractor? {
-        val pathParamAnnotation = parameter.findAnnotation<PathParam>()
-        val parameterName = pathParamAnnotation?.value ?: parameter.name
+    override fun create(function: KFunction<*>, parameter: KParameter,
+            pathPattern: PathPattern): ParameterExtractor? {
+        val pathParamAnnotation = parameter.findAnnotation<PathParam>() ?: return null
+        val parameterName = when (pathParamAnnotation.value.isBlank()) {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/extractors/PathPatternParameterExtractor.kt]: Is this an if clause disguised as a when clause?

File: misk/src/main/kotlin/misk/web/marshal/Marshaller.kt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+package misk.web.marshal
+
+import misk.Action
+import misk.web.ResponseBody
+import okhttp3.MediaType
+
+interface Marshaller<in T> {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/Marshaller.kt]: I like this a lot

File: misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package misk.web.mediatype
+
+import okhttp3.MediaType
+
+object MediaTypes {
+    const val APPLICATION_JSON = "application/json;charset=utf-8"
+    const val TEXT_PLAIN_UTF8 = "text/plain;charset=utf-8"
+}
+
+val String.asMediaType: MediaType get() = MediaType.parse(this)!!
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt]: lazy ?

File: misk/src/main/kotlin/misk/util/Checks.kt (hunk 1 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/util/Checks.kt]: Wasn't aware of require. Spiff

File: misk/src/main/kotlin/misk/web/extractors/PathPatternParameterExtractor.kt (hunk 4 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/web/extractors/PathPatternParameterExtractor.kt]: Yeah, stylistically it seems like we prefer `when` over `if` evaluating as an expression? I've mostly seen our code doing `when`s even with just two conditions. I'm cool changing this to if though, it's more correct.

File: misk/src/main/kotlin/misk/web/extractors/ParameterExtractor.kt [hunk 7]
```diff
@@ -17,9 +18,10 @@ interface ParameterExtractor {
...
          * examples.
          */
-        fun create(parameter: KParameter, pathPattern: PathPattern): ParameterExtractor?
+        fun create(function: KFunction<*>, parameter: KParameter,
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/web/extractors/ParameterExtractor.kt]: nit: one parameter per line if it doesn't fit on a single line.

File: misk/src/main/kotlin/misk/web/extractors/RequestBodyParameterExtractor.kt [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+        return unmarshaller.unmarshal(request.body)
+    }
+
+    class Factory @Inject internal constructor(
+            @JvmSuppressWildcards val unmarshallerFactories: List<Unmarshaller.Factory>
+    ) : ParameterExtractor.Factory {
+        override fun create(
+                function: KFunction<*>,
+                parameter: KParameter,
+                pathPattern: PathPattern): ParameterExtractor? {
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/web/extractors/RequestBodyParameterExtractor.kt]: nit: return type on a newline.

File: misk/src/main/kotlin/misk/web/marshal/Generic.kt [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+
+    object ToByteArray : Unmarshaller {
+        override fun unmarshal(source: BufferedSource): Any? = source.readByteArray()
+    }
+
+    object ToBufferedSource : Unmarshaller {
+        override fun unmarshal(source: BufferedSource): Any? = source
+    }
+
+    object ToByteString : Unmarshaller {
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/web/marshal/Generic.kt]: Doesn't seem like these are used outside this class, can they be private?

File: misk/src/main/kotlin/misk/web/marshal/Marshaller.kt [hunk 10]
```diff
@@ -0,0 +1,14 @@
...
+    fun contentType() : MediaType?
```
Comment (JGulbronson) [inline misk/src/main/kotlin/misk/web/marshal/Marshaller.kt]: When would we marshal to a null `MediaType`?

File: misk/src/test/kotlin/misk/cloud/gcp/security/keys/GcpKeyServiceTest.kt [hunk 11]
```diff
@@ -46,13 +46,17 @@ internal class GcpKeyServiceTest {
         assertThat(plainText.utf8()).isEqualTo("decrypted")
     }
 
-    @Test(expected = IllegalArgumentException::class)
+    @Test
     fun invalidEncryptKey() {
-        keyManager.encrypt("unknown_key", ByteString.encodeUtf8("should fail"))
+        assertThrows(IllegalArgumentException::class.java) {
```
Comment (JGulbronson) [inline misk/src/test/kotlin/misk/cloud/gcp/security/keys/GcpKeyServiceTest.kt]: This file seems to only have Junit 4->5 changes. Should probably be in a separate PR?

File: misk/src/test/kotlin/misk/metrics/MetricsTest.kt [hunk 12]
```diff
@@ -36,9 +36,9 @@ class MetricsTest {
 
         metrics.gauge("my_gauge", { n.get() })
         assertThat(metrics.gauges).containsKey("my_gauge")
-        assertThat(metrics.gauges["my_gauge"]!!.value).isEqualTo(0)
+        assertThat(metrics.gauges["my_gauge"]!!.value).isEqualTo(0L)
```
Comment (JGulbronson) [inline misk/src/test/kotlin/misk/metrics/MetricsTest.kt]: This file just looks like a 4 -> 5 migration.

File: misk/src/main/kotlin/misk/web/marshal/Generic.kt (hunk 9 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/web/marshal/Generic.kt]: yep

File: exemplar/src/main/kotlin/com/squareup/exemplar/HelloWebPostAction.kt [hunk 13]
```diff
@@ -1,19 +1,20 @@
...
 
 @Singleton
 class HelloWebPostAction : WebAction {
     @Post("/hello/{name}")
-    @JsonResponseBody
-    fun hello(
-        name: String,
-        @JsonRequestBody body: PostBody
-    ): HelloPostResponse {
+    @RequestContentType([MediaTypes.APPLICATION_JSON])
```
[the remark this answers is not in the document]
Comment (mmihic) [reply exemplar/src/main/kotlin/com/squareup/exemplar/HelloWebPostAction.kt]: Can you vararg annotation fields?

File: misk/src/main/kotlin/misk/scope/ActionScope.kt [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+@Singleton
+class ActionScope @Inject internal constructor(
+        // NB(mmihic): ActionScoped depends on ActionScope depends on
+        // on ActionScopedProviders, which might depend on other ActionScopeds. We break
+        // this circular dependency by injecting a map of Provider<ActionScopedProvider>
+        // rather than the map of ActionScopedProvider directly
+        private val providers: @JvmSuppressWildcards Map<Key<*>, Provider<ActionScopedProvider<*>>>
+) : AutoCloseable {
+    companion object {
+        private val tls = ThreadLocal<LinkedHashMap<Key<*>, Any>>()
```
[the remark this answers is not in the document]
Comment (mmihic) [reply misk/src/main/kotlin/misk/scope/ActionScope.kt]: Sure

File: misk/src/test/kotlin/misk/cloud/gcp/security/keys/GcpKeyServiceTest.kt (hunk 11 above)
Comment (mmihic) [reply misk/src/test/kotlin/misk/cloud/gcp/security/keys/GcpKeyServiceTest.kt]: Yeah will pull into their own PR

File: misk/src/main/kotlin/misk/scope/ActionScope.kt (hunk 14 above)
[the remark this answers is not in the document]
Comment (mmihic) [reply misk/src/main/kotlin/misk/scope/ActionScope.kt]: Ah, ok, somehow this PR has picked up all of the rebases that I've done.

File: misk/src/test/kotlin/misk/cloud/gcp/security/keys/GcpKeyServiceTest.kt (hunk 11 above)
Comment (mmihic) [reply misk/src/test/kotlin/misk/cloud/gcp/security/keys/GcpKeyServiceTest.kt]: These are actually manik's changes from another branch, not sure why this PR is picking up rebases.

File: misk/src/test/kotlin/misk/metrics/MetricsTest.kt (hunk 12 above)
Comment (mmihic) [reply misk/src/test/kotlin/misk/metrics/MetricsTest.kt]: See above

File: exemplar/src/main/kotlin/com/squareup/exemplar/HelloWebPostAction.kt (hunk 13 above)
[the remark this answers is not in the document]
Comment (maniksurtani) [reply exemplar/src/main/kotlin/com/squareup/exemplar/HelloWebPostAction.kt]: Yes.

File: misk/src/main/kotlin/misk/web/WebActionModule.kt [hunk 15]
```diff
@@ -42,14 +50,25 @@ class WebActionModule<A : WebAction> private constructor(
...
                 for (annotation in member.annotations) {
                     if (annotation !is Get && annotation !is Post) continue
-                    if (member !is KFunction<*>) throw IllegalArgumentException("expected $member to be a function")
-                    if (result != null) throw IllegalArgumentException("multiple annotated methods in $webActionClass")
+                    if (member !is KFunction<*>) throw IllegalArgumentException(
+                            "expected $member to be a function")
+                    if (result != null) throw IllegalArgumentException(
+                            "multiple annotated methods in $webActionClass")
+
+                    // NB(mmihic): The response media type may be ommitted; in this case only
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/WebActionModule.kt]: neat

File: misk/src/main/kotlin/misk/web/interceptors/MarshallerInterceptor.kt [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+import misk.web.marshal.Marshaller
+import misk.web.mediatype.MediaTypes
+import okhttp3.MediaType
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.KType
+import kotlin.reflect.full.findAnnotation
+
+@Singleton
+internal class MarshallerInterceptor constructor(private val marshaller: Marshaller<Any>)
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/interceptors/MarshallerInterceptor.kt]: love this

File: misk/src/main/kotlin/misk/web/marshal/Generic.kt [hunk 17]
```diff
@@ -0,0 +1,111 @@
...
+import kotlin.reflect.jvm.javaType
+
+/** Handles generic unmarshalling, for cases where the action can accept anything */
+object GenericUnmarshallers {
+    fun canHandle(type: KType) = canHandle(type.javaType)
+    fun canHandle(type: Type) = GENERIC_REQUEST_TYPES.contains(type)
+
+    private val GENERIC_REQUEST_TYPES = setOf(
+            String::class.java,
+            Array<Byte>::class.java,
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/Generic.kt]: I wonder if we should deliberately skip this, since it doesn’t do more than ByteString and ByteString is good

File: misk/src/main/kotlin/misk/web/marshal/Generic.kt [hunk 18]
```diff
@@ -0,0 +1,111 @@
...
+
+    private object ToByteString : Unmarshaller {
+        override fun unmarshal(source: BufferedSource): Any? = source.readByteString()
+    }
+
+    fun into(parameter: KParameter): Unmarshaller {
+        val paramType = parameter.type
+        return when (paramType.typeLiteral().rawType) {
+            String::class.java -> ToString
+            Array<Byte>::class.java -> ToByteArray
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/Generic.kt]: off topic: it’ll be really hard to explain to future developers why `Array<Byte>` is a legit raw type and `List<Byte>` is not. No action to take here.

File: misk/src/main/kotlin/misk/web/marshal/Generic.kt [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+            BufferedSource::class.java -> ToBufferedSource
+            ByteString::class.java -> ToByteString
+            else -> throw IllegalArgumentException("no generic unmarshaller for ${parameter.type}")
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/Generic.kt]: should this return null here and let the caller throw?

File: misk/src/main/kotlin/misk/web/marshal/Json.kt [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+        override fun writeTo(sink: BufferedSink) {
+            adapter.toJson(sink, o)
+        }
+    }
+
+    class Factory @Inject internal constructor(val moshi: Moshi) : Marshaller.Factory {
+        override fun create(mediaType: MediaType, type: KType): Marshaller<Any>? {
+            if (mediaType.type() != MediaTypes.APPLICATION_JSON_MEDIA_TYPE.type() ||
+                    mediaType.subtype() != MediaTypes.APPLICATION_JSON_MEDIA_TYPE.subtype())
+                return null
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/Json.kt]: nit: braces ?

File: misk/src/main/kotlin/misk/web/marshal/MarshallerModule.kt [hunk 21]
```diff
@@ -0,0 +1,15 @@
...
+import misk.inject.addMultibinderBinding
+import kotlin.reflect.KClass
+
+class MarshallerModule<T : Marshaller.Factory>(val kclass: KClass<T>) : KAbstractModule() {
+    override fun configure() {
+        binder().addMultibinderBinding<Marshaller.Factory>().to(kclass.java)
+    }
+
+    companion object {
+        inline fun <reified T : Marshaller.Factory> create() = MarshallerModule(T::class)
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/MarshallerModule.kt]: neat

File: misk/src/main/kotlin/misk/web/marshal/PlainText.kt [hunk 22]
```diff
@@ -0,0 +1,54 @@
...
+    override fun contentType() = MediaTypes.TEXT_PLAIN_UTF8_MEDIA_TYPE
+    override fun responseBody(o: Any) = object : ResponseBody {
+        override fun writeTo(sink: BufferedSink) {
+            sink.writeString(o.toString(), Charsets.UTF_8)
+        }
+    }
+
+    class Factory : Marshaller.Factory {
+        override fun create(mediaType: MediaType, type: KType): Marshaller<Any>? {
+            if (mediaType.type() != MediaTypes.TEXT_PLAIN_UTF8_MEDIA_TYPE.type() ||
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/PlainText.kt]: we need a typeAndSubtypeEquals ?

File: misk/src/main/kotlin/misk/web/marshal/PlainText.kt [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+    object ToByteString : Unmarshaller {
+        override fun unmarshal(source: BufferedSource) = source.readByteString()
+    }
+
+    class Factory : Unmarshaller.Factory {
+        override fun create(mediaType: MediaType, type: KType): Unmarshaller? {
+            if (mediaType.type() != MediaTypes.TEXT_PLAIN_UTF8_MEDIA_TYPE.type() ||
+                    mediaType.subtype() != MediaTypes.TEXT_PLAIN_UTF8_MEDIA_TYPE.subtype()) return null
+
+            if (GenericUnmarshallers.canHandle(type)) return null
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/marshal/PlainText.kt]: is this the wrong place for this? Should the thing calling `create()` make this decision?

File: misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt [hunk 24]
```diff
@@ -0,0 +1,118 @@
...
+/** An RFC-2616 media range */
+data class MediaRange(
+        val type: String,
+        val subtype: String,
+        val charset: Charset? = null,
+        val qualityFactor: Double = 1.0,
+        val parameters: Map<String, String> = mapOf(),
+        val extensions: Map<String, String> = mapOf()
+) {
+    val wildcardCount = when (type) {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt]: this is a strange expression!

File: misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt [hunk 25]
```diff
@@ -0,0 +1,118 @@
...
+
+    fun matcher(mediaType: MediaType): Matcher? {
+        val typeMatches = type == mediaType.type() || type == WILDCARD
+        val subtypeMatches = subtype == mediaType.subtype() || subtype == WILDCARD
+        if (!typeMatches || !subtypeMatches) {
+            return null
+        }
+
+        if (charset == null || mediaType.charset() == null) {
+            // One or the other doesn't specify a charset, so we can't match it
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt]: comment suggests _not_ matching, but code suggests matching?

Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt]: ahhh, now I see the boolean

File: misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt [hunk 26]
```diff
@@ -0,0 +1,118 @@
...
+    data class Matcher(val mediaRange: MediaRange, val matchesCharset: Boolean = false)
+
+    companion object {
+        val WILDCARD = "*"
+        val ALL_MEDIA = MediaRange(WILDCARD, WILDCARD)
+
+        fun parse(s: String): MediaRange {
+            val typeParametersAndExtensions = s.split(';')
+            val typeParts = typeParametersAndExtensions[0].split('/')
+            require(typeParts.size == 2) { "$s is not a valid media range" }
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt]: annoyingly, this is an unchecked exception. I think that’s right but callers may need to be careful

Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt]: (though Kotlin doesn’t have checked exceptions . . . I guess I mean that callers will need to catch it)

File: misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt [hunk 27]
```diff
@@ -0,0 +1,118 @@
...
+            val parametersAndExtensions = typeParametersAndExtensions.drop(1).map {
+                parseNameValue(it)
+            }
+            val parameters = LinkedHashMap<String, String>()
+            val extensions = LinkedHashMap<String, String>()
+            var charset: Charset? = null
+            var qualityFactor = 1.0
+            var inParameters = true
+
+            for (p in parametersAndExtensions) {
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/mediatype/MediaRange.kt]: neat. The way q is special is new to me

File: misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt [hunk 28]
```diff
@@ -0,0 +1,17 @@
+package misk.web.mediatype
+
+import okhttp3.MediaType
+
+object MediaTypes {
+    const val APPLICATION_JSON = "application/json;charset=utf-8"
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt]: internal or private ?

File: misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt [hunk 29]
```diff
@@ -0,0 +1,17 @@
...
+    val APPLICATION_JSON_MEDIA_TYPE = APPLICATION_JSON.asMediaType
+
+    const val TEXT_PLAIN_UTF8 = "text/plain;charset=utf-8"
+    val TEXT_PLAIN_UTF8_MEDIA_TYPE = TEXT_PLAIN_UTF8.asMediaType
+
+    const val ALL = "*/*"
+    val ALL_MEDIA_TYPE = ALL.asMediaType
+}
+
+val String.asMediaType: MediaType get() = MediaType.parse(this)!!
```
Comment (swankjesse) [inline misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt]: I think the `as` prefix reads like a function, but this is defined as an extension property. Hm.

File: misk/src/test/kotlin/misk/web/ContentBasedDispatchTest.kt [hunk 30]
```diff
@@ -0,0 +1,133 @@
...
+                .isEqualTo("*/json my friend")
+    }
+
+    @Test
+    fun postArbitraryExpectArbitrary() {
+        val responseContent = post(weirdTextMediaType, "my friend", weirdTextMediaType).source()
+        assertThat(responseContent.readUtf8()).isEqualTo("*/* my friend")
+    }
+
+    class TestModule : KAbstractModule() {
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/ContentBasedDispatchTest.kt]: NEAT

File: misk/src/test/kotlin/misk/web/ContentBasedDispatchTest.kt [hunk 31]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    class PostAnythingReturnAnything : WebAction {
+        @Post("/hello")
+        fun hello(@misk.web.RequestBody message: String) = "*/* $message"
+    }
+
+    private fun post(contentType: MediaType, content: String,
+            acceptedMediaType: MediaType? = null): okhttp3.ResponseBody {
+        val httpClient = OkHttpClient()
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/ContentBasedDispatchTest.kt]: we gotta somehow DI these. Each call to this constructor is slow ’cause it creates its own threadpools

File: misk/src/test/kotlin/misk/web/ContentBasedDispatchTest.kt [hunk 32]
```diff
@@ -0,0 +1,133 @@
...
+    private @Inject lateinit var moshi: Moshi
+    private @Inject lateinit var jettyService: JettyService
+    private val packetJsonAdapter get() = moshi.adapter(Packet::class.java)
+
+    @Test
+    fun postJsonExpectJson() {
+        val requestContent = packetJsonAdapter.toJson(Packet("my friend"))
+        val responseContent = post(jsonMediaType, requestContent, jsonMediaType).source()
+        assertThat(packetJsonAdapter.fromJson(responseContent)!!.message)
+                .isEqualTo("json/json my friend")
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/ContentBasedDispatchTest.kt]: first read through I read the `/` as the media type separator, like `application/json`. A comma would be a little less surprising, or even `"request=json response=json my friend"`

File: misk/src/test/kotlin/misk/web/marshal/JsonRequestTest.kt [hunk 33]
```diff
@@ -0,0 +1,114 @@
...
+
+    class PassAsByteString : WebAction {
+        @Inject lateinit var moshi: Moshi
+        private val packetJsonAdapter get() = moshi.adapter(Packet::class.java)
+
+        @Post("/as-byte-string")
+        @RequestContentType(MediaTypes.APPLICATION_JSON)
+        @ResponseContentType(MediaTypes.APPLICATION_JSON)
+        fun call(@RequestBody encodedPacket: ByteString): Packet {
+            val source = Okio.buffer(Okio.source(ByteArrayInputStream(encodedPacket.toByteArray())))
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/marshal/JsonRequestTest.kt]: or `Buffer().write(encodedPacket)`

File: misk/src/test/kotlin/misk/web/marshal/JsonResponseTest.kt [hunk 34]
```diff
@@ -0,0 +1,160 @@
...
+    class ReturnAsString : WebAction {
+        @Get("/response/as-string")
+        @ResponseContentType(MediaTypes.APPLICATION_JSON)
+        fun call() = "{\"message\":\"as-string\"}"
+    }
+
+    class ReturnAsByteString : WebAction {
+        @Get("/response/as-byte-string")
+        @ResponseContentType(MediaTypes.APPLICATION_JSON)
+        fun call(): ByteString = ByteString.encodeUtf8("{\"message\":\"as-byte-string\"}")
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/marshal/JsonResponseTest.kt]: don’t forget `"""` strings

File: misk/src/test/kotlin/misk/web/marshal/JsonResponseTest.kt [hunk 35]
```diff
@@ -0,0 +1,160 @@
...
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okio.BufferedSink
+import okio.ByteString
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.Test
+import javax.inject.Inject
+
+@ActionTest(startService = true)
+internal class JsonResponseTest {
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/marshal/JsonResponseTest.kt]: Love how these tests work

File: misk/src/test/kotlin/misk/web/marshal/PlainTextRequestTest.kt [hunk 36]
```diff
@@ -0,0 +1,80 @@
...
+        @RequestContentType(MediaTypes.TEXT_PLAIN_UTF8)
+        @ResponseContentType(MediaTypes.TEXT_PLAIN_UTF8)
+        fun call(@RequestBody message: String): String = "$message as-string"
+    }
+
+    class PassAsByteString : WebAction {
+        @Post("/as-byte-string")
+        @RequestContentType(MediaTypes.TEXT_PLAIN_UTF8)
+        @ResponseContentType(MediaTypes.TEXT_PLAIN_UTF8)
+        fun call(@RequestBody messageBytes: ByteString): String  = "${messageBytes.utf8()} as-byte-string"
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/marshal/PlainTextRequestTest.kt]: does anyone want to pass strings as ByteStrings?

File: misk/src/test/kotlin/misk/web/mediatype/MediaRangeTest.kt [hunk 37]
```diff
@@ -0,0 +1,215 @@
...
+    }
+
+    @Test
+    fun parseWithParameters() {
+        val mediaRange = MediaRange.parse("text/html;level = 1 ; strict = true")
+        assertThat(mediaRange.type).isEqualTo("text")
+        assertThat(mediaRange.subtype).isEqualTo("html")
+        assertThat(mediaRange.charset).isNull()
+        assertThat(mediaRange.extensions).isEmpty()
+        assertThat(mediaRange.parameters).containsExactly("level" to "1", "strict" to "true")
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/mediatype/MediaRangeTest.kt]: neat

File: misk/src/test/kotlin/misk/web/mediatype/MediaRangeTest.kt [hunk 38]
```diff
@@ -0,0 +1,215 @@
...
+    @Test
+    fun blankSubType() {
+        assertThrows(IllegalArgumentException::class.java) {
+            MediaRange.parse("text/")
+        }
+    }
+
+    @Test
+    fun noTypeSubType() {
+        assertThrows(IllegalArgumentException::class.java) {
```
Comment (swankjesse) [inline misk/src/test/kotlin/misk/web/mediatype/MediaRangeTest.kt]: YASS

File: misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt (hunk 28 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt]: It's intended for use by applications, so neither?

File: misk/src/main/kotlin/misk/web/marshal/PlainText.kt (hunk 23 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/web/marshal/PlainText.kt]: I don't think so - I think we want to allow each media type handler to decide whether it wants to defer to generic handling.

File: misk/src/main/kotlin/misk/web/marshal/Json.kt (hunk 20 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/web/marshal/Json.kt]: It seems like that's not our coding style in Kotlin? (I've seen multiple places where we do single line if's without braces)

File: misk/src/main/kotlin/misk/web/marshal/Generic.kt (hunk 19 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/web/marshal/Generic.kt]: yeah that's reasonable

File: misk/src/main/kotlin/misk/web/marshal/Generic.kt (hunk 17 above)
Comment (mmihic) [reply misk/src/main/kotlin/misk/web/marshal/Generic.kt]: I'm cool with that

File: misk/src/main/kotlin/misk/web/marshal/Json.kt (hunk 20 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/marshal/Json.kt]: works for me. I think in Java we do single-line and not single statement.

File: misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt (hunk 28 above)
Comment (swankjesse) [reply misk/src/main/kotlin/misk/web/mediatype/MediaTypes.kt]: Ahh, cause the annotation can’t take a MediaType, it needs to take a string. Yep
